### ISSUE TITLE
FR: Make usable for dev environment

### ISSUE THREAD

**Issue by ryanwwest** [zephyros-dev/docker-koreader#20]
I've been trying to set up KOReader plus X11 forwarding on a headless Ubuntu server for development (and connecting from my Macbook with M1) and kept running into errors in both their Docker version and running it direct (and KOReader on M1 ARM chips doesn't play well either..). Then this docker image just immediately worked with zero hassle, VNC seems easier than X11.

My ultimate goal, though, is a dev environment through VS Code and a way to test the application UI while changing files. Is there a way to let this build KOReader from a volumed-in `./koreader.git` repo and rebuild with latest file changes whenever you restart the container?

**Comment by zephyros-dev** [comment]
You can try the VSCode devcontainer that works similarly to this image [here](https://github.com/zephyros-dev/koreader/tree/devcontainer/.devcontainer). Once I finish polishing the settings and the docs I will create a PR at the Koreader repository. For now you can add the devcontainer folder above into your local Koreader repository and use the devcontainer feature of VSCode to setup the dev environment. For more info regarding the devcontainer you can check out this [link](https://code.visualstudio.com/docs/devcontainers/create-dev-container)
Once the devcontainer is running, you can access the Desktop via the forwarded port feature
![image](https://github.com/zephyros-dev/docker-koreader/assets/67867662/900b563c-6856-4e65-8967-16469abefe56)
The rest of the development steps is identical to the Koreader documentations (.e.g: you can run `./kodev run` in the terminal and see the Koreader instance on the browser)

**Comment by ryanwwest** [comment]
Thanks @zephyros-dev. On a related note, would you know how to best monitor KOReader's logs as it runs (after enabling debug verbose logging within the UI)? That appears to only pipe to the process's stdout and not a file, but the docker container logs are now only displaying the VNC client, not the underlying app.

**Comment by ryanwwest** [comment]
The solution wasn't too hard, you can just run `pkill -9 koreader; koreader` within the docker-koreader container and that will restart KOReader but with logs going into stdout now. If you just exit KOReader yourself it will be auto-restarted (without stdout) by the VNC server, but this way I guess VNC still accepts you doing a manual restart before it can.

Of course I don't think there's a full koreader repo of code here so I'm not sure about `kodev` - it looks like a binary alone is used? But for plugin-only development, this works.

**Comment by zephyros-dev** [comment]
> Of course I don't think there's a full koreader repo of code here so I'm not sure about `kodev` - it looks like a binary alone is used? But for plugin-only development, this works.

In my original post I was using the devcontainer inside the orignal `koreader` repository (which contains the original source code along with the kodev binary) rather than using this `docker-koreader` repository. I haven't updated it in a while though so not sure if it is still working, if I have some time I will take a look at that again.
For the logs, after you run the process using the integrated terminal of VSCode inside the devcontainer, you can see the stdout logs from there.

**Comment by ryanwwest** [comment]
Yes I've been viewing the logs within vs code's integrated terminal as you suggested. I don't think I got the container json file working but am otherwise connected.
